dot/core/messages.go

```diff
@@ -41,56 +78,29 @@ func (s *Service) HandleTransactionMessage(peerID peer.ID, msg *network.Transact
 		return false, err
 	}
 
+	allTxsAreValid := true
 	for _, tx := range txs {
-		err = func() error {
-			s.storageState.Lock()
-			defer s.storageState.Unlock()
-
-			ts, err := s.storageState.TrieState(&head.StateRoot)
-			if err != nil {
-				return err
-			}
-
-			rt.SetContextStorage(ts)
-
-			// validate each transaction
-			externalExt := types.Extrinsic(append([]byte{byte(types.TxnExternal)}, tx...))
-			val, err := rt.ValidateTransaction(externalExt)
-			if err != nil {
-				if errors.Is(err, runtime.ErrInvalidTransaction) {
-					s.net.ReportPeer(peerset.ReputationChange{
-						Value:  peerset.BadTransactionValue,
-						Reason: peerset.BadTransactionReason,
-					}, peerID)
-				}
-				logger.Debugf("failed to validate transaction: %s", err)
-				return nil
-			}
-
-			// create new valid transaction
-			vtx := transaction.NewValidTransaction(tx, val)
-
-			// push to the transaction queue of BABE session
-			hash := s.transactionState.AddToPool(vtx)
-			logger.Tracef("added transaction with hash %s to pool", hash)
+		validity, isValidTxn, err := s.validateTransaction(peerID, head, rt, tx)
+		if err != nil {
+			return false, fmt.Errorf("failed validating transaction for peerID %s: %w", peerID, err)
+		}
 
+		if !isValidTxn {
+			allTxsAreValid = false
```

Comment:
should we break here?

Reply:
I dont believe so. I think we want to still check for other valid transactions to add to the pool. We just need to know not to report the peer for good behavior if they have invalid txns i do believe